### PR TITLE
feat(dashmate): doctor diagnostic

### PR DESCRIPTION
**Pull request by shumkov**
## Issue being fixed or feature implemented

We want the dashmate doctor command not just to report but to analyse reports and identify basic issues. Sine the list of possible checks is endless, we want to implement only basic ones for the first iteration.

## What was done?

- Spilt doctor command into two commands:
  - `dashmate doctor` to collect and analyse samples. 
  - `dashmate doctor report` to create an archive with
- Added an option to analyse already created archive 
- Analyse collected data:
  - Docker installation
  - Configuration
    -  Gateway admin is disabled while metrics are enabled
    - Platform Node ID
    - SSL certificate (file, self-signed, zerossl)
    - Core P2P port
    - Gateway HTTP port
    - Tenderdash P2P port
  - Core
    - Core is syncing
    - PoSe
  - Platform
    - Tenderdash is syncing
  - Service containers
    - Container is not started
    - Container exited with error code
    - Container is OOMKilled
  - System resources
    - CPU cores
    - CPU speed
    - CPU load
    - RAM total
    - RAM free
    - Swap
    - Disk space
    - Disk IO wait time
    - Open file descriptors
 - Memorise docker installation check

## How Has This Been Tested?

- Run commands locally
- Simulate different conditions
- Added unit tests

## Breaking Changes

None

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [x] I have made corresponding changes to the documentation if needed

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

**Comment by shumkov** [comment]
https://github.com/orgs/dashpay/projects/3/views/2?pane=issue&itemId=78731831

File: packages/dashmate/src/doctor/analyse/analyseCoreFactory.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+      problems.push(problem);
+    }
+
+    // PoSe
+    if (samples?.getDashmateConfig()?.get('core.masternode.enable')) {
+      const masternodeStatus = samples.getServiceInfo('core', 'masternodeStatus');
+
+      const { description, solution, severity } = {
+        WAITING_FOR_PROTX: {
```
Comment (shumkov) [inline packages/dashmate/src/doctor/analyse/analyseCoreFactory.js:38]: @thephez please help me with this

**Review by pshenmic — APPROVED** [review]
utACK

**Comment by qwizzie** [comment]
@shumkov is 'Container is not started' (personally i would use 'Container is not starting'), the same as a container stuck in 'restarting' ? Because that (restarting) i have noticed with MNO's having an issue with their Evonode, when they report the outcome of a dashmate status services command. If it is not the same, then maybe it needs to be added ?

File: packages/dashmate/src/doctor/verifySystemRequirementsFactory.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    const totalMemory = dockerSystemInfo?.MemTotal ?? memory?.total;
+
+    if (Number.isInteger(totalMemory)) {
+      const totalMemoryGb = totalMemory / (1024 ** 3); // Convert to GB
+
+      if (totalMemoryGb < MINIMUM_RAM) {
+        const problem = new Problem(
+          `${totalMemoryGb.toFixed(2)}GB RAM detected. At least ${MINIMUM_RAM}GB is required`,
+          `Consider to upgrade RAM to make sure the node can provide required services
+to the network in time and will not get PoSe banned`,
```
Comment (thephez) [inline packages/dashmate/src/doctor/verifySystemRequirementsFactory.js:66]:
```suggestion
          `Consider upgrading RAM to make sure the node can provide timely responses
for required network services and avoid Proof-of-Service bans`,
```

File: packages/dashmate/src/doctor/verifySystemRequirementsFactory.js [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+      console.warn('Can\'t get memory information');
+    }
+
+    // CPU speed
+    if (cpu && Number.isFinite(cpu.speed) && cpu.speed !== 0) {
+      if (cpu.speed < MINIMUM_CPU_FREQUENCY) {
+        const problem = new Problem(
+          `${cpu.speed.toFixed(1)}GHz CPU frequency detected. At least ${MINIMUM_CPU_FREQUENCY}GHz is required`,
+          `Consider to upgrade CPU to make sure the node can provide required services
+to the network in time and will not get PoSe banned`,
```
Comment (thephez) [inline packages/dashmate/src/doctor/verifySystemRequirementsFactory.js:83]:
```suggestion
          `Consider upgrading CPUs to make sure the node can provide timely responses
for required network services and avoid Proof-of-Service bans`,
```

File: packages/dashmate/src/doctor/verifySystemRequirementsFactory.js [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+    // Check swap information
+    if (memory && Number.isInteger(memory.swaptotal)) {
+      const swapTotalGb = (memory.swaptotal / (1024 ** 3)); // Convert bytes to GB
+
+      if (swapTotalGb < 2) {
+        const problem = new Problem(
+          `Swap space is ${swapTotalGb.toFixed(2)}GB. 2GB is recommended`,
+          `Consider to enable SWAP to make sure the node can provide required services
+to the network in time and will not get PoSe banned`,
```
Comment (thephez) [inline packages/dashmate/src/doctor/verifySystemRequirementsFactory.js:102]:
```suggestion
          `Swap space is ${swapTotalGb.toFixed(2)}GB. 2GB is recommended`,
          `Consider enabling SWAP to make sure the node can provide timely responses
for required network services and avoid Proof-of-Service bans`,
```

File: packages/dashmate/src/doctor/verifySystemRequirementsFactory.js [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+    // Get disk usage info
+    if (diskSpace) {
+      const availableDiskSpace = diskSpace.available / (1024 ** 3); // Convert to GB
+
+      if (availableDiskSpace < MINIMUM_DISK_SPACE) {
+        const problem = new Problem(
+          `${availableDiskSpace.toFixed(2)}GB available disk space detected. At least ${MINIMUM_DISK_SPACE}GB is required`,
+          `Consider to increase disk space to make sure the node can provide required services
+to the network in time and will not get PoSe banned`,
```
Comment (thephez) [inline packages/dashmate/src/doctor/verifySystemRequirementsFactory.js:118]:
```suggestion
          `${availableDiskSpace.toFixed(2)}GB of available disk space detected. At least ${MINIMUM_DISK_SPACE}GB is required`,
          `Consider increasing disk space to make sure the node can provide timely responses
for required network services and avoid Proof-of-Service bans`,
```

**Review by thephez — COMMENTED** [review]
@shumkov Took longer than I hoped :sweat_smile: Here is an initial look over everything...

File: packages/dashmate/src/doctor/analyse/analyseCoreFactory.js [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+          severity: SEVERITY.HIGH,
+        },
+        POSE_BANNED: {
+          description: 'The masternode has been banned due to failing Proof-of-Service checks.',
+          solution: chalk`Submit a ProUpServTx transaction to unban your masternode and ensure
+it meets all network requirements.`,
+          severity: SEVERITY.HIGH,
+        },
+        REMOVED: {
+          description: 'The masternode has been removed from the network masternode list.',
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/dashmate/src/doctor/analyse/analyseCoreFactory.js:50]: Not gonna work this way. It needs to be escaped.

[the remark this answers is not in the document]
Comment (thephez) [reply packages/dashmate/src/doctor/analyse/analyseCoreFactory.js:50]: Was trying to pay attention to those, but I saw somewhere it was not escaped and got confused. Updated the suggestion

File: packages/dashmate/src/doctor/analyse/analyseCoreFactory.js [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        },
+        OPERATOR_KEY_CHANGED: {
+          description: 'The operator key for the masternode has been changed.',
+          solution: chalk`Update the masternode configuration with the new operator key
+using {bold.cyanBright dashmate config set core.masternode.operatorKey <operatorKey>}.`,
+          severity: SEVERITY.HIGH,
+        },
+        PROTX_IP_CHANGED: {
+          description: 'The IP address registered in the ProTx has changed.',
+          solution: chalk`Update your masternode's configuration with the new IP address.`,
```
Comment (thephez) [inline packages/dashmate/src/doctor/analyse/analyseCoreFactory.js:62]: This one probably needs to be escaped too?
```suggestion
          solution: chalk`Update your masternode\'s configuration with the new IP address.`,
```

File: packages/dashmate/src/doctor/analyse/analyseCoreFactory.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/dashmate/src/doctor/analyse/analyseCoreFactory.js:50]: You don't need to escape ' if you use `.

File: packages/dashmate/src/commands/doctor/index.js [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+    const problems = ctx.prescription.getOrderedProblems();
+    if (problems.length === 0) {
+      // eslint-disable-next-line no-console
+      console.log(chalk`\n  The doctor didn't find any problems with your node.
+
+  If issues still persist, please use {bold.cyanBright dashmate doctor report} to create an archive
+  of already collected data for further investigation.
+
+  You can use it to analyze your node condition yourself or send it to the Dash Core Group ({underline.cyanBright support@dash.org}).`);
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/dashmate/src/commands/doctor/index.js:89]: No need this case. You need to escape only if you quite string with the same symbol.

File: packages/dashmate/src/commands/doctor/index.js [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+
+    const prescriptionString = chalk`\n  ${problemsCount} found:
+
+${problemsString}
+
+
+  Use {bold.cyanBright dashmate doctor report} to create an archive
+  of already collected data for further investigation.
+
+  You can use it to analyze your node condition yourself or send it to the Dash Core Group ({underline.cyanBright support@dash.org}).`;
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/dashmate/src/commands/doctor/index.js:137]: No need to escape in this case

File: packages/dashmate/src/commands/doctor/report.js [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+  - System information
+  - The node configuration
+  - Service logs, metrics and status
+
+  Collected data will not contain any private information which is already not available publicly.
+  All sensitive data like private keys or passwords is obfuscated.
+
+  The archive will compressed with TAR/GZIP and placed to {bold.cyanBright ${process.cwd()}}
+  You can use it to analyze your node condition yourself or send it to the Dash Core Group ({underline.cyanBright support@dash.org}).\n`,
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/dashmate/src/commands/doctor/report.js:55]: No need to escape in this case

File: packages/dashmate/src/commands/doctor/index.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (thephez) [reply packages/dashmate/src/commands/doctor/index.js:89]: Makes sense now :+1: Updated these

**Review by thephez — DISMISSED** [review]
Doc ACK :+1:

**Review by QuantumExplorer — APPROVED** [review]
Not reviewed, Approved based on other reviewers.

**Pull request merged**